src/Build.UnitTests/TestData/ProjectInstanceTestObjects.cs

```diff
@@ -105,14 +105,14 @@ public static ProjectOnErrorInstance CreateTargetOnError(int? counter = null)
 
         public static ProjectTaskInstance CreateTargetTask(
             int? counter = null,
-            IDictionary<string, Tuple<string, ElementLocation>> parameters = null,
+            IDictionary<string, (string, ElementLocation)> parameters = null,
```

Comment:
Does putting ValueTuples in a Dictionary mean they get copied when they're added in or accessed and stored elsewhere? I don't know how Dictionary works very well, but I think this is the point:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L652

Reply:
Think of a Dictionary conceptually like any other collection (basically array).

For ValueTypes, the structs are inlined, there are no separate objects anywhere. Reads and writes copy the entire struct back and forth.

For reference types, the objects exist elsewhere, with 8 (or 16) bytes overhead per object (not including the payload), and the array stores references (4 or 8 bytes). So in this case having a reference type is very wasteful.

Reply:
Also keep in mind that with "inline" structs that are lined up in memory you reap the benefits of cache locality. The CPU prefetches a chunk of memory into the cache and blazes through it. For reference types it has potentially to fetch a page for each reference because it is stored elsewhere in the heap.

Reply:
I was with you up to the last line. The reference type does one allocation (and the value type does 0), but when reading or writing, the reference type only has to copy a small number of bytes, so everything after that is almost free, whereas the value type has to copy a lot of information back and forth, which takes CPU time. Having one allocation sounds better to me. Can you explain further?

Reply:
Re: locality, if they're stored at roughly the same time, I imagine they'd have fairly good locality either way.

Reply:
Remember I'm talking in general statements, one needs to profile for each specific case. But the problem is that you don't have just one allocation, you have potentially hundreds of thousands, and they all add GC pressure and huge memory overhead. Allocations in .NET is a big problem, so I'd rather opt into less allocations by default than optimize struct copying. Yes, structs will be copied back and forth but honestly in my experience I never remember that being a perf problem. It's always the allocations that kill you. CPU copies memory around all the time and it's insanely fast. Copying 16 bytes around is nothing compared to the work done when doubling an array to the next power of two when growing a list, and then effectively doing O(n^2) copies, and we never even talk about those costs.

And soon with ref structs etc. we'll even be able to avoid copying the structs, we'll be just passing a reference to the struct that stays in place.

Reply:
@Forgind you've nerd-sniped me and I made an actual benchmark :) https://github.com/KirillOsenkov/Benchmarks/blob/fbc760fec7f6b310c223e50f9f564ff0135917c7/Benchmarks/Tests/DictionaryOfValueVsReferenceTypes.cs#L10-L11

```
|         Method |     Mean |   Error |  StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |---------:|--------:|--------:|-------:|------:|------:|----------:|
|     ValueTypes | 303.1 ns | 4.62 ns | 4.32 ns | 0.0572 |     - |     - |     240 B |
| ReferenceTypes | 303.4 ns | 3.58 ns | 3.17 ns | 0.0658 |     - |     - |     276 B |
```

Basically there's no perf overhead from struct copying in this case, just 36 bytes less allocations ;) Free allocation reduction!

Reply:
Thanks! I wonder if part of why it's so efficient is that the items in the structs are tiny, so copying back and forth is no worse than copying a few pointers in the form of "what's in the struct."

That (and your comment about allocations) convinces me this is a good change. Don't feel the need to check this, but I'd be curious about it if it were a massive struct, i.e., a struct with a massive non-reference type, vs. a class. I'm not sure what that would be, though, since you can't use a class or a string or any other reference types, and a struct has minimal (0?) overhead. Maybe a struct with a large number of things in it would work, but then you start to wonder if anyone has ever done that.

...and then it would become fully irrelevant with ref structs.